NTumbleBit/ClassicTumbler/Client/StateMachinesExecutor.cs

```diff
@@ -51,20 +51,22 @@ protected override void StartCore(CancellationToken cancellationToken)
                         if(lastCycle != cycle.Start)
                         {
                             // Only start a new cycle if there are sufficient wallet funds
-                            Money walletBalance = this.Runtime.Services.WalletService.GetBalance();
-                            Money minimumBalance = this.Runtime.TumblerParameters.Denomination + this.Runtime.TumblerParameters.Fee;
-
-                            if (walletBalance >= minimumBalance)
+                            if (Runtime.HasEnoughFundsForCycle())
                             {
                                 lastCycle = cycle.Start;
-                                Logs.Client.LogInformation("New Cycle: " + cycle.Start);
+                                Logs.Client.LogInformation("New Cycle: {0}", cycle.Start);
                                 PaymentStateMachine.State state = GetPaymentStateMachineState(cycle);
                                 if (state == null)
                                 {
                                     var stateMachine = new PaymentStateMachine(Runtime, null);
                                     stateMachine.NeedSave = true;
                                     Save(stateMachine, cycle.Start);
                                 }
+                            } else
+                            {
+                                Logs.Client.LogInformation("Tumbling finished; there is no enough funds to sustain another tumbling cycle.");
+                                Stopped();
```

Comment:
I am not so sure we should be doing this here; there will possibly be other cycles still executing that need to run to completion. Doesn't this terminate the state machine prematurely?

Reply:
That's a very good call @zeptin. I think that the Stopped() method might actually stop all cycles. I will investigate how to handle that in the way that the last cycle will stop the process. Let's keep this PR open for now.